### PR TITLE
Add TF image classification example script

### PR DESCRIPTION
**Pull request by amyeroberts**
# What does this PR do?

Adds the TF equivalent for the PyTorch image classification example script. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Thanks a lot for working on this!
It looks like your branch is a bit old and does not contain some fixes made to make sure the example tests run when an example is modified (you can see the test examples are not running here 😅 ). Could you try a rebase on main?

**Comment by sgugger** [comment]
Still no good, I scanned the test fetcher and found a potential bug. COuld you add
```py
            elif f.startswith("examples/tensorflow"):
                test_files_to_run.append("examples/flax/test_tensorflow_examples.py")
```
at the line 562 of `utils/test_fetcher.py` (in-between PyTorch and Flax)? I think that's what causing the issue of the tests not running.

**Comment by amyeroberts** [comment]
> It looks like your branch is a bit old and does not contain some fixes made to make sure the example tests run when an example is modified (you can see the test examples are not running here 😅 ). Could you try a rebase on main?

@sgugger I've rebased from upstream main and force pushed again. If I run `git log --oneline` I can see these changes are applied on top of the tip of main. 
```
270bfb056 (HEAD -> add-examples-tf-image-classification, origin/add-examples-tf-image-classification) Add tests
a2256258b Fix up
1a1594cb8 Update requirements
b6a2f1ef9 TF image classification script
9ccea7acb (upstream/main, main) Fix some doctests after PR 15775 (#20036)
a639ea9e8 Add **kwargs (#20037)
ec6878f6c Now supporting pathlike in pipelines too. (#20030)
aa39967b2 reorganize glossary (#20010)
305e8718b Show installed libraries and their versions in CI jobs (#20026)
```

The examples still aren't running and I can't see in the diff where I could be overriding this 😅 I'm sure there's something I'm overlooking. I'll keep digging but let me know if there's something else I should be doing. Sorry to bother.

File: examples/tensorflow/image-classification/run_image_classification.py [hunk 1]
```diff
@@ -0,0 +1,567 @@
...
+            j = np.random.randint(0, width - w + 1)
+            return image[i : i + h, j : j + w, :]
+
+    # Fallback to central crop
+    in_ratio = float(width) / float(height)
+    w = width if in_ratio < min(ratio) else int(round(height * max(ratio)))
+    h = height if in_ratio > max(ratio) else int(round(width / min(ratio)))
+    i = (height - h) // 2
+    j = (width - w) // 2
+    return image[i : i + h, j : j + w, :]
```
Comment (Rocketknight1) [inline examples/tensorflow/image-classification/run_image_classification.py:207]: I know this is from torchvision and not your idea, but trying 10 random rolls and then giving up rather than figuring out how to constrain your randomness so it outputs valid solutions the first time is quite funny.

File: examples/tensorflow/image-classification/run_image_classification.py [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+    else:
+        image_size = (feature_extractor.size["height"], feature_extractor.size["width"])
+
+    def _train_transforms(image):
+        img_size = image_size
+        image = tf.keras.utils.img_to_array(image)
+        image = random_resized_crop(image, size=img_size)
+        image = tf.image.random_flip_left_right(image)
+        image /= 255.0
+        image = (image - feature_extractor.image_mean) / feature_extractor.image_std
```
Comment (Rocketknight1) [inline examples/tensorflow/image-classification/run_image_classification.py:347]: I haven't looked at our image models too much, but I'm guessing that the scaling factors `feature_extractor.image_mean` and `feature_extractor.image_std` were also computed after division by 255? It seems odd to normalize the scale twice like this.

**Review by Rocketknight1 — APPROVED** [review]
This looks really clean overall! Have you tested that you get similar accuracy after fine-tuning as the PT example? I know the CI test checks this, but the tolerances on quick small tests like that have to be quite wide, so they can miss subtler performance degradation.

File: examples/tensorflow/image-classification/run_image_classification.py (hunk 1 above)
Comment (Rocketknight1) [reply examples/tensorflow/image-classification/run_image_classification.py:207]: Although on reflection I suppose doing it this way gives you an unbiased sample from the space of valid target_area + aspect ratio pairs, and you'd need to handle things quite carefully to make that the case if you constrained the random rolls to always be valid.

**Comment by Rocketknight1** [comment]
Hey @amyeroberts, is this PR still going ahead? It looked almost ready!

File: examples/tensorflow/image-classification/run_image_classification.py (hunk 2 above)
Comment (amyeroberts) [reply examples/tensorflow/image-classification/run_image_classification.py:347]: Exactly. It's pretty standard to have the values this way in vision - most of our `image_mean` and `image_std` values are constants from e.g. imagenet. I think it comes down to being able to use floats - if I found the mean on my R pixels was 134.7 how should I take it away from the pixel values when they're still uint8?

**Comment by amyeroberts** [comment]
@Rocketknight1 Yes - sorry, this fell down my priority list for a bit. Code is all ready to go - I was trying to find models that make the tests run quickly c.f. [this comment](https://github.com/huggingface/transformers/pull/19956#discussion_r1012991408).

**Comment by Rocketknight1** [comment]
@amyeroberts Ah, that makes sense! It's totally okay to upload your own super-mini model and use that - it doesn't really matter if the accuracy is bad, the test will just let us detect if the outputs from this model class change suddenly

File: examples/tensorflow/image-classification/README.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+Internally, the script will use the [`ImageFolder`](https://huggingface.co/docs/datasets/v2.0.0/en/image_process#imagefolder) feature which will automatically turn the folders into 🤗 Dataset objects.
+
+##### 💡 The above will split the train dir into training and evaluation sets
+  - To control the split amount, use the `--train_val_split` flag.
+  - To provide your own validation split in its own directory, you can pass the `--validation_dir <path-to-val-root>` flag.
+
+#### Upload your data to the hub, as a (possibly private) repo
+
+It's very easy (and convenient) to upload your image dataset to the hub using the [`ImageFolder`](https://huggingface.co/docs/datasets/v2.0.0/en/image_process#imagefolder) feature available in 🤗 Datasets. Simply do the following:
```
Comment (MKhalusova) [inline examples/tensorflow/image-classification/README.md:102]: Minor nitpick: I would generally avoid saying that's something is "very easy" because this can sound condescending to beginners. You can also skip "simply" in "Simply do the following" without losing any meaning.

File: examples/tensorflow/image-classification/README.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
+<!---
+Copyright 2021 The HuggingFace Team. All rights reserved.
```
Comment (amyeroberts) [inline examples/tensorflow/image-classification/README.md:2]:
```suggestion
Copyright 2023 The HuggingFace Team. All rights reserved.
```

File: examples/tensorflow/image-classification/README.md (hunk 3 above)
Comment (amyeroberts) [reply examples/tensorflow/image-classification/README.md]: Updated!

File: examples/tensorflow/test_tensorflow_examples.py [hunk 5]
```diff
@@ -294,3 +296,27 @@ def test_run_translation(self):
             run_translation.main()
             result = get_results(tmp_dir)
             self.assertGreaterEqual(result["bleu"], 30)
+
+    @slow
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply examples/tensorflow/test_tensorflow_examples.py]: So, this was paused for quite a while because the script was taking ages on the `prepare_tf_dataset` step. Thanks to the sleuthing from @Rocketknight1 and the subsequent to_tf_dataset updates and using `with_format("np")` things move a lot faster. Switching to a `hf-internal-testing` dataset and small model meant it runs < 30s.
